spatial/curve/hilbert.go

```diff
@@ -0,0 +1,241 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package curve
+
+// Hilbert2D is a 2-dimensional Hilbert curve.
+type Hilbert2D struct{ Order int }
+
+// Dims returns the spatial dimensions of the curve, which is {2ᵏ, 2ᵏ}, where k
+// is the order.
+func (h Hilbert2D) Dims() []int { return []int{1 << h.Order, 1 << h.Order} }
+
+// Len returns the length of the curve, which is 2ⁿᵏ, where n is the dimension
+// (2) and k is the order.
+//
+// Len will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Len() int { return 1 << (2 * h.Order) }
+
+func (h Hilbert2D) rot(n int, v []int, d int) {
+	switch d {
+	case 0:
+		swap{0, 1}.do(n, v)
+	case 3:
+		flip{0, 1}.do(n, v)
+	}
+}
+
+// Pos returns the linear position of the spatial coordinate along the curve.
+// Pos modifies v.
+//
+// Pos will overflow on a 32-bit architecture if the order is ≥ 16.
+func (h Hilbert2D) Pos(v []int) int {
+	var d int
+	for n := h.Order - 1; n >= 0; n-- {
+		rx := (v[0] >> n) & 1
+		ry := (v[1] >> n) & 1
+		rd := ry<<1 | (ry ^ rx)
+		d += rd << (2 * n)
+		h.rot(h.Order, v, rd)
+	}
+	return d
+}
+
+// WriteCoordTo writes the spatial coordinates of pos to v.
+func (h Hilbert2D) WriteCoordTo(pos int, v []int) {
+	for n := 0; n < h.Order; n++ {
+		e := pos & 3
+		h.rot(n, v[:], e)
+
+		ry := e >> 1
+		rx := (e>>0 ^ e>>1) & 1
+		v[0] += rx << n
+		v[1] += ry << n
+		pos >>= 2
+	}
+}
+
+// Coord returns the spatial coordinates of pos as a slice.
+func (h Hilbert2D) Coord(pos int) []int {
+	v := make([]int, 2)
+	h.WriteCoordTo(pos, v)
+	return v
+}
```

Comment:
Do we need this? It seems like a helper that suggests inefficient client code.

Reply:
:shrug:. If I were writing code where the performance of `Coord` was unimportant, I'd rather use this signature. WriteCoordTo is kind of awkward.

Reply:
Given that we're removing `Coord`, should the other method stay `WriteCoordTo`?

Reply:
Good point. Maybe `Coord(dst []int, pos int)`? Also, on a survey, maybe we *can* have something like `Coord(dst []int, pos int) []int` which allocates the slice is `dst` is nil. I know that this goes against what I wrote above, but it's consistent both with existing code we have here and with similar functions that are in the standard library.

Reply:
I like `Coord(dst []int, pos int) []int`. As you say that's an established pattern and it addresses both use cases.

Reply:
Done